### PR TITLE
storage/rangefeed: per registration buffer

### LINKED ISSUE #32945: changefeedccl: Prevent slow consumers from blocking rangefeed Processor
Work item to track fixing slow-consumer blocking in the rangefeed processor. This involves installing a per-registration event buffer which the processor writes to, along with a per-registration goroutine which reads from that buffer and outputs to the target stream.

Detailed problem and proposal outlined in #32944.

### PR DESCRIPTION
**Pull request by mrtracy**
Modifies the rangefeed package so that each individual registration
maintains an output buffer which is processed by a dedicated goroutine.
This eliminates the possibility of a slow consumer blocking the
processor from making progress.

Previously, writing events to registrations was synchronous, meaning
that a single slow consumer could cause the publishing process to block.
If publishing was blocked for too long, the processor along with all
registrations were immediately torn down. With this change, a slow
consumer will no longer block its siblings; if a single consumer blocks
for too long, it will enter an "overflow" state and will be torn down as
  soon as its existing buffer is fully processed.

The registry/registration interface has been modified fairly
significantly to accomodate this; the processor itself has seen modest
modification (mostly moving functionality to the registry level and
removing no-longer-needed support for the previous functionality). Some
of the processor tests have been modified to account for the now
asynchronous nature of publishing events to individual streams.

Fixes #32945

Release note: None

### CODE REVIEW

**Review by nvb — COMMENTED** [review]
Aside from a few small questions, this looks really good! Have you tested it with the roachtests yet?

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-:-LVkRqxe-Kx5-NvMRMs2:b-i8oi92)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/processor.go, line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVjr46TDj1-TtB-n2DS:-LVjr46TDj1-TtB-n2DT:b91h3lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6ac8d44e8c219a503bcf255064f1636620266c9/pkg/storage/rangefeed/processor.go#L372)):*
> ```Go
> 	case p.regC <- r:
> 	case <-p.stoppedC:
> 		if catchUpIter != nil {
> ```

Why did this change? Will this cause us to leak iterators if the Processor is stopped?
___
*[pkg/storage/rangefeed/registry.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVjrNJPBmNVR-5Ra_7C:-LVjrNJPBmNVR-5Ra_7D:b9jzdvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L71)):*
> ```Go
> 
> 	mu struct {
> 		*syncutil.Mutex
> ```

`sync.Locker`

Or pass a pointer to a `registration` everywhere and don't make this pointer to a mutex.
___
*[pkg/storage/rangefeed/registry.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVk1-F2BL1p0LmLc7ee:-LVk1-F2BL1p0LmLc7ef:b-a67yyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L150)):*
> ```Go
> // 1. If a catch-up scan is indicated, run one before beginning the proper
> // output loop.
> // 2. After catch-up is complete, Begin reading from the registration buffer
> ```

s/Begin/begin/
___
*[pkg/storage/rangefeed/registry.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVk1EOv6EHl4nBYyuWL:-LVk1EOv6EHl4nBYyuWM:bc2nz3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L187)):*
> ```Go
> 		case <-ctx.Done():
> 			return ctx.Err()
> 		case <-r.stream.Context().Done():
> ```

:thumbsup: I'm happy that we can do this instead of polling all stream contexts.
___
*[pkg/storage/rangefeed/registry.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkRVuiCbVVf0RZ3-YH:-LVkRVuiCbVVf0RZ3-YI:b9o13z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L193)):*
> ```Go
> }
> 
> func (r *registration) runOutputLoop(ctx context.Context, doneCb func(r *registration)) {
> ```

If this is executing synchronously then why take a callback instead of just running the disconnect code after this returns?
___
*[pkg/storage/rangefeed/registry.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkQZiE5hTeE3WOrCpb:-LVkQZiE5hTeE3WOrCpc:b-7msk0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L206)):*
> ```Go
> // runCatchupScan starts a catchup scan which will output entries for all
> // recorded changes in the replica that are newer than the catchupTimeStamp.
> // This creates a single engine iterator immediately when this method
> ```

Is this sentence true?
___
*[pkg/storage/rangefeed/registry.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkQj-A8hta-y90VB8T:-LVkQj-A8hta-y90VB8U:b-epmwss) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L274)):*
> ```Go
> }
> 
> func (r *registration) waitForCaughtUp() error {
> ```

This is just used in tests, right? Can we move it to a `_test.go` file so that it isn't part of the `registration` type's method set outside of tests?
___
*[pkg/storage/rangefeed/registry.go, line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkQo3lEo6b4GTTPJi2:-LVkQo3m66pksqtvbr8D:b-fx0stz) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L432)):*
> ```Go
> }
> 
> func (reg *registry) waitForCaughtUp() error {
> ```

Same question about moving this to a `_test.go` file.

**Review by mrtracy — COMMENTED** [review]
TFTR! Have not run these in roachtests yet, but we've only recently started running roachtests with rangefeed enabled at all.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-:-LWAX9BG41a_EWln7-Zh:b5nmr19)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/processor.go, line 372 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVjr46TDj1-TtB-n2DS:-LWAUQb83wtrt3NaHYIQ:bq533iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6ac8d44e8c219a503bcf255064f1636620266c9/pkg/storage/rangefeed/processor.go#L372)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why did this change? Will this cause us to leak iterators if the Processor is stopped?
</blockquote></details>

I'm pretty sure it would, responsibility for closing the catchupIter was with the registration, at this point it gets leaked if the processor closes while registration is in progress.

I'm not sure anymore if this is one of the leaked routines I saw in a test explicitly, I think I just caught it visually.
___
*[pkg/storage/rangefeed/registry.go, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVjrNJPBmNVR-5Ra_7C:-LWAVHCo5OkFjzF1jlYi:bobt5gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L71)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`sync.Locker`

Or pass a pointer to a `registration` everywhere and don't make this pointer to a mutex.
</blockquote></details>

Done, used Locker.
___
*[pkg/storage/rangefeed/registry.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVk1-F2BL1p0LmLc7ee:-LWAVOg50gffXcShGk8i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L150)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/Begin/begin/
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/registry.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVk1EOv6EHl4nBYyuWL:-LWAVPoQ0fE4Klcf68PS:bsoq2vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L187)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

:thumbsup: I'm happy that we can do this instead of polling all stream contexts.
</blockquote></details>

Agreed, It's quite a bit easier to understand it from the perspective of a single registration, even with the elaborate shutdown dance for the dedicated goroutine.
___
*[pkg/storage/rangefeed/registry.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkRVuiCbVVf0RZ3-YH:-LWAWS3z7F4E4471CKJN:b-iuowyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L193)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If this is executing synchronously then why take a callback instead of just running the disconnect code after this returns?
</blockquote></details>

Ah, good eye. I think I originally had this being directly run as the async task, but then I created that outer task in the processor which could just be doing this itself, but then I missed the shortcut.
___
*[pkg/storage/rangefeed/registry.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkQZiE5hTeE3WOrCpb:-LWAWmQyBvLOs7cODFbZ:bton5s8) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L206)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this sentence true?
</blockquote></details>

Nope, made some significant changes here and forgot to update the comment.  Fixed.
___
*[pkg/storage/rangefeed/registry.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkQj-A8hta-y90VB8T:-LWAX3LR2cWkhiHP3i2h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L274)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is just used in tests, right? Can we move it to a `_test.go` file so that it isn't part of the `registration` type's method set outside of tests?
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/registry.go, line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LVkQo3lEo6b4GTTPJi2:-LWAX5cpEGYPV8tVddHi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/registry.go#L432)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same question about moving this to a `_test.go` file.
</blockquote></details>

Done.

**Review by danhhz — COMMENTED** [review]
:lgtm: I still can't say I'm an expert on this code -- I suspect I'll have to make a modification or two myself before I'm 100% comfortable with it -- but everything here makes sense to me. Given that nathan seems to have signed off on the high-level and he's away until after you leave, I'd say feel free to merge this when you're happy with it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-:-LWHvhWdDRGJjSTtgri6:b7kro3e)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/rangefeed/processor.go, line 329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LViC4S3EpAQnT_UbKCi:-LViC4S42dDMb3fB34Xq:b-wnnjju) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/processor.go#L329)):*
> ```Go
> // provided an error when the registration closes.
> //
> // The provided engine is used to instantiate "catch-up" iterators for this
> ```

nit: it's back to an iterator
___
*[pkg/storage/rangefeed/registry.go, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LWHuH6a2KdzImJXfBc-:-LWHuH6bBZKtX673pmMY:b-jcyltz) ([raw file](https://github.com/cockroachdb/cockroach/blob/44862d0d664d70252f1f8c45aaba3d6ce2cf5dae/pkg/storage/rangefeed/registry.go#L156)):*
> ```Go
> func (r *registration) outputLoop(ctx context.Context) error {
> 	// If the registration has a catch-up scan,
> 	if !r.catchupTimestamp.IsEmpty() {
> ```

it seems more natural to me for this to be triggered by a non-nil `catchupIter`

**Review by mrtracy — COMMENTED** [review]
I had to change the way that "waitForCatchup" works in order to avoid a race condition; that functionality now runs through the processor. As a result, the waitForCatchup() methods cannot be moved to the _test package, so I have clearly marked them for test-only usage in other ways.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-:-LWIlZT81fJeqd7PoM48:bv0g153)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/rangefeed/processor.go, line 329 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LViC4S3EpAQnT_UbKCi:-LWIlUB72Vruhcj0BG_5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/545bd291617d823c21c5d53be95d34a4ef9cd589/pkg/storage/rangefeed/processor.go#L329)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: it's back to an iterator
</blockquote></details>

Done.
___
*[pkg/storage/rangefeed/registry.go, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/33557#-LWHuH6a2KdzImJXfBc-:-LWIl0P23DeQ2Q5Amxl-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44862d0d664d70252f1f8c45aaba3d6ce2cf5dae/pkg/storage/rangefeed/registry.go#L156)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

it seems more natural to me for this to be triggered by a non-nil `catchupIter`
</blockquote></details>

Done.

**Review by danhhz — APPROVED** [review]

**Comment by mrtracy** [comment]
bors r=danhhz

**Pull request merged**
